### PR TITLE
Add a pattern and a pass to rewrite linalg_ext::tile to linalg_ext::in_parallel

### PR DESCRIPTION
**Pull request by nicolasvasilache**

### CODE REVIEW

File: include/Dialects/LinalgExt/LinalgExtOps.td [hunk 1]
```diff
@@ -223,9 +223,15 @@ def LinalgExt_InParallelOp : LinalgExt_PureOp<"in_parallel", [
   let parser = [{ return ::parse$cppClass(parser, result); }];
   let verifier = [{ return ::verify(*this); }];
 
+  // The efault builder does not add the proper body BBargs, roll our own.
```
Comment (apaszke) [inline include/Dialects/LinalgExt/LinalgExtOps.td:226]: nit: typo in efault

File: lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    Value totalSize =
+        rewriter.create<tensor::DimOp>(loc, tileOp.outs().front(), zero);
+    Value step = tileOp.tile_sizes();
+    assert(step.getType().isa<IndexType>() && "NYI: not an index type");
+
+    AffineBuilder ab(rewriter, loc);
+    AffineExpr i, j, M;
+    bindDims(rewriter.getContext(), i, j);
+    bindSymbols(rewriter.getContext(), M);
+    Value sizePerThread = ab.ceil(Tie{i, totalSize}, Tie{M, step});
```
Comment (apaszke) [inline lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp:102]: It's not really `sizePerThread`, right? It's more like `numThreads` or `numTiles`. The previous name was a little confusing for me

File: lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    // Construct the op without a body builder: we need to clone the ops in the
+    // body explicitly after having access to the new bbArgs.
+    // As a consequence, `ensureTerminator` is not called and the body has no
+    // terminator.
+    linalg_ext::InParallelOp inParallelOp =
+        rewriter.create<linalg_ext::InParallelOp>(loc, tileOp->getResultTypes(),
+                                                  sizePerThread);
+    inParallelOp.ensureTerminator(inParallelOp.region(), rewriter, loc);
```
Comment (apaszke) [inline lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp:111]: Do we still need this? I think that with the new bodyless builder this should be unnecessary.

File: lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    SmallVector<Value> bbArgsTranslated{offset, size};
+    llvm::append_range(bbArgsTranslated, implicitSubtensorExtracts);
+    rewriter.mergeBlockBefore(&tileOp.region().front(),
+                              inParallelOp.getBody()->getTerminator(),
+                              bbArgsTranslated);
+
+    // tileOp's terminator is not the terminator, insert explicit subset_insert
+    // ops and feed them to a new scf.yield terminator that we can now add.
+    auto performConcurrentlyOp =
+        cast<PerformConcurrentlyOp>(&inParallelOp.getBody()->back());
```
Comment (apaszke) [inline lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp:143]: Should we add a helper to `InParallelOp` that returns the terminator already cast to `PerformConcurrentlyOp`? We do that in a whole bunch of places

File: lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+    // TODO: getBody() -> Block*
```
Comment (apaszke) [inline lib/Dialects/LinalgExt/Transforms/TileToInParallel.cpp:145]: ??

File: test/LinalgExt/tile_to_in_parallel.mlir [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+// CHECK-DAG: #[[$CEIL_MAP:.*]] = affine_map<(d0)[s0] -> (d0 ceildiv s0)>
+// CHECK-DAG: #[[$MUL_MAP:.*]] = affine_map<(d0)[s0] -> (d0 * s0)>
+// CHECK-DAG: #[[$SUB_MAP:.*]] = affine_map<(d0, d1) -> (d0 - d1)>
+// CHECK-DAG: #[[$ID1_MAP:.*]] = affine_map<(d0) -> (d0)>
+// CHECK-DAG: #[[$ID2_MAP:.*]] = affine_map<(d0, d1) -> (d0, d1)>
+
+// CHECK-LABEL: func @static_tile
+//  CHECK-SAME:   %[[CHUNK_SIZE:[0-9a-z]+]]: index
+//  CHECK-SAME:   %[[IN:[0-9a-z]+]]: tensor<?xf32>
+//  CHECK-SAME:   %[[OUT:[0-9a-z]+]]: tensor<?xf32>
```
Comment (apaszke) [inline test/LinalgExt/tile_to_in_parallel.mlir:12]: CHECK-SAME means that they should appear on the same line?

File: test/LinalgExt/tile_to_in_parallel.mlir [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+func @static_tile(%chunk_size: index, %in: tensor<?xf32>, %out: tensor<?xf32>) -> (tensor<?xf32>) {
+  %c0 = arith.constant 0: index
+
+  // CHECK: %[[M:.*]] = tensor.dim %{{.*}}, %{{.*}} : tensor<?xf32>
+  // CHECK: %[[CEIL:.*]] = affine.apply #[[$CEIL_MAP]](%[[M]])[%[[CHUNK_SIZE]]]
+  // CHECK: linalg_ext.in_parallel %[[CEIL]] -> (tensor<?xf32>) {
+  %0 = linalg_ext.tile %chunk_size outs(%out: tensor<?xf32>) -> (tensor<?xf32>) {
+
+  // CHECK: ^bb0(%[[TIDX:.*]]: index
```
Comment (apaszke) [inline test/LinalgExt/tile_to_in_parallel.mlir:21]: Check that there are no more args? I think you just need `):` at the end

**Review by apaszke — APPROVED** [review]
Very nice! A few minor points

File: test/LinalgExt/tile_to_in_parallel.mlir (hunk 6 above)
Comment (nicolasvasilache) [reply test/LinalgExt/tile_to_in_parallel.mlir:12]: yes

**Pull request merged**